Lombiq.EmailClient/Services/ImapEmailClient.cs

```diff
@@ -0,0 +1,237 @@
+            },
+            Content = new EmailContent
+            {
+                Body = new EmailBody
+                {
+                    Body = message.TextBody ?? message.HtmlBody,
+                    IsHtml = !string.IsNullOrEmpty(message.HtmlBody),
+                },
+            },
+        };
+
+        emailMessage.Header.To.AddRange(CreateEmailAddresses(message.To.Mailboxes));
+        emailMessage.Header.Cc.AddRange(CreateEmailAddresses(message.Cc.Mailboxes));
+        emailMessage.Header.Bcc.AddRange(CreateEmailAddresses(message.Bcc.Mailboxes));
+
+        emailMessage.Content.Attachments.AddRange(message.Attachments
+            .OfType<MimePart>()
+            .Select(attachment => new AttachmentMetadata
+            {
+                FileName = attachment.FileName,
+                MimeType = attachment.ContentType.MimeType,
+                Size = attachment.Content.Stream?.Length ?? 0,
+            }));
+
+        return emailMessage;
+    }
+
+    private static IEnumerable<EmailAddress> CreateEmailAddresses(IEnumerable<MailboxAddress> mailboxes) =>
+        mailboxes.Select(m => new EmailAddress { DisplayName = m.Name, Address = m.Address });
+
+    private async Task<IMailFolder> OpenFolderAsync(string folderName = null)
+    {
+        var folder = string.IsNullOrEmpty(folderName)
+            ? _imapClient.Inbox
+            : await _imapClient.GetFolderAsync(folderName);
+        await folder.OpenAsync(FolderAccess.ReadOnly);
+
+        return folder;
+    }
+
+    private async Task InitializeImapClientAndConnectIfNeededAsync()
+    {
+        if (_imapClient?.IsConnected ?? false) return;
+
+        _imapClient = new ImapClient();
+        await _imapClient.ConnectAsync(_imapSettings.Host, _imapSettings.Port, _imapSettings.UseSsl);
+
+        var username = _imapSettings.RequireAuthentication ? _imapSettings.Username : "user";
+        var password = _imapSettings.RequireAuthentication ? _imapSettings.Password : string.Empty;
+        await _imapClient.AuthenticateAsync(username, password);
+    }
+
+    private static void ValidateImapProtocol(EmailMessage emailMessage)
+    {
+        if (emailMessage.Metadata.Protocol != Protocols.Imap)
+        {
+            throw new InvalidOperationException("The email message protocol is not IMAP.");
+        }
+    }
+
+    private static async Task<MemoryStream> SaveAttachmentToMemoryAsync(MimeEntity attachment)
+    {
+        var stream = new MemoryStream();
+        if (attachment is MessagePart rfc822)
+        {
+            await rfc822.Message.WriteToAsync(stream);
+        }
+        else
+        {
+            var part = (MimePart)attachment;
+
+            await part.Content.DecodeToAsync(stream);
+        }
+
+        stream.Position = 0;
+
+        return stream;
+    }
+
+    private bool IsLocalhost() =>
+        _imapSettings.Host == "127.0.0.1" ||
+        _imapSettings.Host.Contains("localhost", StringComparison.OrdinalIgnoreCase);
```

Comment:
Apparently there is a canonical way to check if an `IPAddress` is localhost/loopback, you just have to convert the host name into IP first like this: `Dns.GetHostAddresses(host).Any(IPAddress.IsLoopback)`